From d62d30773decf3cd44cbd2a4edc8f0daf7c0ca5b Mon Sep 17 00:00:00 2001
From: Chong Zhang <chz@google.com>
Date: Thu, 14 Jun 2018 12:28:32 -0700
Subject: [PATCH 02/80] stagefright: send EOS to encoder for thumbnail
 extraction

Decoders might hold on to decoded frames for certain streams,
send EOS after last sample so that we get all output frames.

bug: 110116757
bug: 110226430

Test: CTS HeifWriterTest; MediaMetadataRetrieverTest
Change-Id: I0f8e7f1607857297325594adad8ca4fbff16e5e9
---
 media/libstagefright/FrameDecoder.cpp | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/media/libstagefright/FrameDecoder.cpp b/media/libstagefright/FrameDecoder.cpp
index 29a219f5f..6e94517a3 100644
--- a/media/libstagefright/FrameDecoder.cpp
+++ b/media/libstagefright/FrameDecoder.cpp
@@ -301,10 +301,13 @@ status_t FrameDecoder::extractInternal() {
             err = mSource->read(&mediaBuffer, &mReadOptions);
             mReadOptions.clearSeekTo();
             if (err != OK) {
-                ALOGW("Input Error or EOS");
                 mHaveMoreInputs = false;
                 if (!mFirstSample && err == ERROR_END_OF_STREAM) {
+                    (void)mDecoder->queueInputBuffer(
+                            index, 0, 0, 0, MediaCodec::BUFFER_FLAG_EOS);
                     err = OK;
+                } else {
+                    ALOGW("Input Error: err=%d", err);
                 }
                 break;
             }
-- 
2.17.1

